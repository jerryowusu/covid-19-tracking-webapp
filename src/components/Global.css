:root {
    --pink-light: #fe7f9c;
    --pink-dark: #f64a8a;
    --pink-darker: #ce4176;
    --blue-light: #5687e4;
    --white: #fff;
}

.header {
    display: flex;
    justify-content: space-around;
    background-color: var(--pink-light);
    color: white;
}

.search-form {
    display: flex;
    justify-content: space-around;
    background-color: var(--pink-dark);
    color: white;
    height: 2rem;
    padding: 0.5rem;
    font-size: 1.1rem
}

.search-input {
    width: 14rem;
    border: none;
    border-radius: 0.5rem;
    font-size: 1.3rem;
    padding: 0.5rem;
    outline: none;
}

.all-cards {
    display: grid;
    grid-template-columns: repeat(2, 1fr);
    background-color: var(--pink-light);
    color: white;
    margin-top: -0.15rem;
    padding: 0.1rem;
    font-size: 0.7rem
}

a:nth-child(4n) {
    background-color: var(--pink-dark);
}

a:nth-child(4n+1) {
    background-color: var(--pink-dark);
}

a:nth-child(4n+2) {
    background-color: var(--pink-darker);
}

a:nth-child(4n+3) {
    background-color: var(--pink-darker);
}

.country-card {
    height: 18vh;
    color: white;
}

.country-images {
    display: flex;
    padding: 1rem;
    height: 50%;
}

.country-card .flag {
   width: 40%;
   height: 70%;
   opacity: 0.7;
}

.text {
    margin-top: -2.0rem;
    text-align: right;
    font-size: 1rem;
    padding: 0.65rem;
}

a {
    text-decoration: none;
}