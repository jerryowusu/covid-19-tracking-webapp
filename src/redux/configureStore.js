import { createStore, applyMiddleware, combineReducers } from 'redux';
import thunk from 'redux-thunk';
import logger from 'redux-logger';
import covidReducer from './covidData';

const reducer = combineReducers({
  covidReducer,
});

const store = createStore(
  reducer,
  (applyMiddleware(logger, thunk)),
);

export default store;
