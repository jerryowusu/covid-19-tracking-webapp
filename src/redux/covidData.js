const GET_DATA = 'GET_DATA';

const initialState = [];

export const getData = (payload) => ({
  type: GET_DATA,
  payload,
});

const covidReducer = (state = initialState, action) => {
  switch (action.type) {
    case GET_DATA:
      return [...state, ...action.payload];

    default: {
      return state;
    }
  }
};

export default covidReducer;
