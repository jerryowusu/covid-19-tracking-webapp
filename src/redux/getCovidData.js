const getDataFromAPI = async () => {
  const info = [];
  const response = await fetch('https://disease.sh/v3/covid-19/countries');
  const dataFromAPI = await response.json();

  dataFromAPI.map(({ countryInfo: { _id: id, flag }, ...data }) => {
    const covidData = {
      continent: data.continent,
      country: data.country,
      country_id: id,
      country_flag: flag,
      total_cases: data.cases,
      total_deaths: data.deaths,
      total_recovered: data.recovered,
      total_active: data.active,
      total_tests: data.tests,
      population: data.population,
      todays_cases: data.todayCases,
      todays_deaths: data.todayDeaths,
      todays_recovered: data.todayRecovered,
    };

    return info.push(covidData);
  });

  return info;
};

export default getDataFromAPI;
