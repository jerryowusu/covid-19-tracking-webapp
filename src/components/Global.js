import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { BsArrowRightCircle } from 'react-icons/bs';
import { getData } from '../redux/covidData';
import getDataFromAPI from '../redux/getCovidData';
import worldMap from '../img/worldmap.png';
import './Global.css';

const Global = () => {
  const covidData = useSelector((store) => store.covidReducer);
  const dispatch = useDispatch();

  useEffect(() => {
    if (covidData.length === 0) {
      getDataFromAPI()
        .then((response) => dispatch(getData(response)));
    }
  }, []);

  const [findCountry, setFindCountry] = useState('');

  const filteredCountries = [];

  if (covidData.length > 0 && findCountry) {
    covidData.filter((each) => {
      if (each.country.toLowerCase().includes(findCountry)) {
        filteredCountries.push(each);
      }
      return filteredCountries;
    });
  }

  const selectCountry = (element) => setFindCountry(element.target.value);

  return (
    <section>
      <header className="header">
        <img className="img" src={worldMap} alt="world-map" width="60%" height="60%" />
        <div className="wrapper">
          <h1 className="header-title">World</h1>
          <h3 className="header-number">444,638,071</h3>
        </div>
      </header>
      <form className="search-form">
        <span className="form-span">Infections by Country</span>
        <input className="search-input" name="search" type="text" placeholder="country" onChange={selectCountry} />
      </form>

      <ul className="all-cards">
        {
      filteredCountries && filteredCountries.map((country) => (
        <Link className="country-lists" key={country.country} to={{ pathname: `/country/${country.country}` }}>
          <div className="country-card">
            <div className="country-images">
              <img src={country.country_flag} alt="flag" className="flag" />
              <BsArrowRightCircle className="infos-circle" />
            </div>
            <h1 className="text">{country.country}</h1>
            <p className="text">{country.total_cases}</p>
          </div>
        </Link>
      ))
    }
        {
      filteredCountries.length === 0 && covidData.map((country) => (
        <Link className="country-lists" key={country.country} to={{ pathname: `/country/${country.country}` }}>
          <div className="country-card">
            <div className="country-images">
              <img src={country.country_flag} alt="flag" className="flag" />
              <BsArrowRightCircle className="infos-circle" />
            </div>
            <h1 className="text">{country.country}</h1>
            <p className="text">{country.total_cases}</p>
          </div>
        </Link>
      ))
    }
      </ul>
    </section>
  );
};

export default Global;
