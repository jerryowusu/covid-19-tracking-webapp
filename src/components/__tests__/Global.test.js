import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import React from 'react';
import { Provider } from 'react-redux';
import { combineReducers, createStore, applyMiddleware } from 'redux';
import renderer from 'react-test-renderer';
import thunk from 'redux-thunk';
import Global from '../Global';
import covidReducer from '../../redux/covidData';

const mockCovidData = [
  {
    country_flag: 'flag-1',
    country: 'country 1',
    population: '121321',
    total_cases: '45',
    total_recovered: '1',
    total_active: '54',
    total_deaths: '5',
  },
  {
    country_flag: 'flag-1',
    country: 'country 1',
    population: '121321',
    total_cases: '45',
    total_recovered: '1',
    total_active: '54',
    total_deaths: '5',
  },
  {
    country_flag: 'flag-1',
    country: 'country 1',
    population: '121321',
    total_cases: '45',
    total_recovered: '1',
    total_active: '54',
    total_deaths: '5',
  },
];

const rootReducer = combineReducers({
  covidReducer,
});

const store = createStore(
  rootReducer,
  applyMiddleware(thunk),
);

describe('Global component tests:', () => {
  it('Global component renders correctly', () => {
    const globalPage = renderer.create(
      <Provider store={store}>
        <Global country={mockCovidData} />
      </Provider>,
    ).toJSON();
    expect(globalPage).toMatchSnapshot();
  });

  it('Number of countries displayed should be none', () => {
    const { container } = render(
      <Provider store={store}>
        <Global country={mockCovidData} />
      </Provider>,
    );
    const page = container.querySelector('.all-cards');
    expect(page.childElementCount).toBe(0);
  });
});
