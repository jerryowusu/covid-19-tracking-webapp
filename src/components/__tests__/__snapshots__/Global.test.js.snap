// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global component tests: Global component renders correctly 1`] = `
<section>
  <header
    className="header"
  >
    <img
      alt="world-map"
      className="img"
      height="60%"
      src="worldmap.png"
      width="60%"
    />
    <div
      className="wrapper"
    >
      <h1
        className="header-title"
      >
        World
      </h1>
      <h3
        className="header-number"
      >
        444,638,071
      </h3>
    </div>
  </header>
  <form
    className="search-form"
  >
    <span
      className="form-span"
    >
      Infections by Country
    </span>
    <input
      className="search-input"
      name="search"
      onChange={[Function]}
      placeholder="country"
      type="text"
    />
  </form>
  <ul
    className="all-cards"
  />
</section>
`;
